sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/Utility.java

```diff
@@ -287,25 +287,28 @@ public static OffsetDateTime parseDate(String dateString) {
                             String.format("Request body emitted %d bytes, less than the expected %d bytes.",
                                 currentTotalLength[0], length), currentTotalLength[0], length));
                     }
-                    return ByteBuffer.wrap(cache);
-                }))
-                .doOnComplete(() -> {
-                    try {
-                        if (data.read() != -1) {
-                            long totalLength = currentTotalLength[0] + data.available();
-                            throw LOGGER.logExceptionAsError(new UnexpectedLengthException(
-                                String.format("Request body emitted %d bytes, more than the expected %d bytes.",
-                                    totalLength, length), totalLength, length));
-                        } else if (currentTotalLength[0] > length) {
-                            throw LOGGER.logExceptionAsError(new IllegalStateException(
-                                String.format("Read more data than was requested. Size of data read: %d. Size of data"
-                                    + " requested: %d", currentTotalLength[0], length)));
+
+                    // Validate that stream isn't longer.
+                    if (currentTotalLength[0] >= length) {
+                        try {
+                            if (data.read() != -1) {
```

Comment:
It's fine to call data.read here since we validate above right? So anything in the start/middle of the stream won't execute this code

Reply:
Just validating my understanding

Reply:
correct, this part was in onComplete before so happend after stream reading (but who knows when exactly).
Now it's executed when last byte buffer is computed so it became predictable.
The top `if` that checks length makes sure that we do this after we reached `length` provided as input.